### DISCUSSION TITLE
Cuda out of memory

### DISCUSSION THREAD

**Discussion by dmandair** [Lightning-AI/pytorch-lightning#11582] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I'm getting the dreaded 'Cuda out of memory error'. I originally had batch sizes of ~200 and reduced this to 175 and this seemed to fix the issue - the problem is I now am intermittently getting the same error on this decreased batch size without changing anything in my code (except for one line in my validation_epoch_end function, bolded below). Am I doing something wrong or that can be obviously fixed? These training batch sizes seem small for what I've typically otherwise been able to handle with other models. My lightning module is below, the underlying models are essentially a convnet encoder followed by an attention mechanism - I'd love any help with this if possible.

Full error:
            RuntimeError: CUDA error: out of memory CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect. For debugging consider passing CUDA_LAUNCH_BLOCKING=1.

Pytorch version: 1.10.1
Cuda 11.3

    class MilaNet(pl.LightningModule):
        def __init__(self, train_data_csv,
                     val_data_csv,
                     attn_temp=None,
                     num_heads=2,
                     attn_depth=2,
                     tile_thresh=150,
                     attention=2,
                     feat_extract=4,
                     qkv_dim=256,
                     num_classes=5,
                     learning_rate=1e-5,
                     optimizer='Adam',
                     feat_dim=256,
                     weight_decay=0,
                     dropout=0.1,
                     hidden_dim=256):

            super(MilaNet, self).__init__()
            torch.manual_seed(0)
            optim_choices = ['Adam', 'Adagrad', 'RMSprop', 'SGD']
            assert optimizer in optim_choices, f'ERROR: Use optimizer in {optim_choices}'
            self.train_data_csv = train_data_csv
            self.val_data_csv = val_data_csv
    
            self.num_heads = num_heads
            self.tile_thresh = tile_thresh
            self.lr = learning_rate
            self.optimizer = optimizer
            self.wd = weight_decay
            self.num_classes = num_classes
            if feat_extract == 1:
                self.featurizer = Featurizer_resnet18(channels=3)
                self.feat_dim = 1024
            elif feat_extract == 2:
                self.featurizer = Featurizer_resnet50(channels=3)
                self.featurizer.fc = nn.Linear(256 * 4, feat_dim)
                self.feat_dim = feat_dim
            elif feat_extract == 3:
                self.featurizer = FeaturizerSimple()
                self.feat_dim = 256
            elif feat_extract == 4:
                self.featurizer = tv_models.resnet18(pretrained=True)
                self.featurizer.fc = nn.Linear(512, feat_dim)
                self.feat_dim = feat_dim
            elif feat_extract == 5:
                self.featurizer = tv_models.resnet18(pretrained=True)
                self.feat_dim = 1000
                num_children = sum([1 for _ in self.featurizer.children()])
                child_counter = 0
                for child in self.featurizer.children():
                    if child_counter >= num_children - 1:
                        for param in child.parameters():
                            param.requires_grad = True
                        child_counter += 1
                    else:
                        for param in child.parameters():
                            param.requires_grad = False
                        child_counter += 1
    
            if attention == 1:
                self.attention = NNAttention(num_heads=num_heads, feat_dim=self.feat_dim,
                                            qkv_dim=qkv_dim, num_classes=num_classes,
                                            dropout=dropout)
            elif attention == 2:
                self.attention = Attention(self.feat_dim, qkv_dim=qkv_dim, num_heads=num_heads,
                                            temp=attn_temp, attn_depth=attn_depth,
                                            num_classes=num_classes, hidden_dim=hidden_dim,
                                            dropout=dropout)
    
            self.val_ds = None
    
        def forward(self, x, batch_indices):
            x = self.featurizer(x)
            out = self.attention(x, batch_indices)
            return out
    
        def loss_function(self, logits, labels):
            return F.cross_entropy(logits, labels)
    
        ###### Training Steps ########
        def training_step(self, batch, batch_idx):
            x, y, batch_indices = batch
            logits = self.forward(x, batch_indices)
            loss = self.loss_function(logits, y)
    #        am_out = torch.argmax(F.softmax(logits, dim=1), dim=1)
            return {'loss': loss, 'pred': logits, 'gt': y}
    
        def training_epoch_end(self, outputs):
            losses = [output['loss'] for output in outputs]
            logits = [output['pred'] for output in outputs]
            y_true = [output['gt'] for output in outputs]
            avg_loss = torch.stack(losses).mean()
            y_pred = torch.argmax(F.softmax(torch.cat(logits, dim=0), dim=1), dim=1).cpu().numpy()
            y_true = torch.cat(y_true).cpu().numpy()
            confusion_mat = sk_metrics.confusion_matrix(y_true, y_pred, labels=range(self.num_classes))
            fig = px.imshow(confusion_mat)
            wandb.log({'Confusion Matrix train': fig})
            train_acc = np.sum(np.equal(y_true, y_pred))/y_true.shape[0]
            self.log('avgerage training loss', avg_loss, logger=True)
            self.log('training accuracy epoch', train_acc, logger=True)
    
        ###### Validation Steps ######
        def validation_step(self, val_batch, batch_idx):
            x, y, batch_indices = val_batch
            logits = self.forward(x, batch_indices)
            loss = self.loss_function(logits, y)
    #        am_out = torch.argmax(F.softmax(logits, dim=1), dim=1)
            self.log('validation loss', loss, logger=True)
            return loss, logits, y
    
        def validation_epoch_end(self, outputs):
            losses = [output[0] for output in outputs]
            logits = [output[1] for output in outputs]
            y_true = [output[2] for output in outputs]
            avg_loss = torch.stack(losses).mean()
            y_pred = torch.argmax(F.softmax(torch.cat(logits, dim=0), dim=1), dim=1).cpu().numpy()
            **scores = F.softmax(torch.cat(logits, dim=0), dim=1)[:,1].detach().cpu().numpy()**
            y_true = torch.cat(y_true).cpu().numpy()
            print('Y_PRED:', y_pred)
            print('Y_TRUE:', y_true)
            print('EQUALS:', np.equal(y_true, y_pred).astype('int'))
            confusion_mat = sk_metrics.confusion_matrix(y_true, y_pred, labels=range(self.num_classes))
            fig = px.imshow(confusion_mat)
            auc = sk_metrics.roc_auc_score(y_true, scores)
            wandb.log({'Confusion Matrix Val': fig, "AUC": auc})
            val_acc = np.sum(np.equal(y_true, y_pred))/y_true.shape[0]
            self.log('average validation loss', avg_loss, logger=True)
            self.log('validation accuracy epoch', val_acc, logger=True)
    
        #### Defining DataLoaders ####
        def train_dataloader(self):
            bags = TissueBagsBulkCV(self.train_data_csv, TILE_PATH, transform = transforms.Compose([
                transforms.ToTensor(),
                transforms.RandomHorizontalFlip(),
                transforms.RandomAffine(0, translate=(0.05,0.05), scale=(0.9,1.1), shear=10),
                transforms.RandomResizedCrop(224)]))
            sampler = TissueBagsSampler(bags.bag_sizes, thresh=self.tile_thresh)
            dl = data_utils.DataLoader(bags, batch_size=None, sampler=sampler)
            return dl
    
        def val_dataloader(self):
            bags = TissueBagsBulkCV(self.val_data_csv, TILE_PATH, transform = transforms.ToTensor())
    #        self.val_ds = bags
            sampler = TissueBagsSampler(bags.bag_sizes, thresh=self.tile_thresh)
            dl = data_utils.DataLoader(bags, batch_size=None, sampler=sampler, shuffle = False)
            return dl
    
        #### Defining Optimizers #####
        def configure_optimizers(self):
            if self.optimizer == 'RMSprop':
                optimizer = torch.optim.RMSprop(self.parameters(), lr=self.lr, weight_decay=self.wd)
            elif self.optimizer == 'Adam':
                optimizer = torch.optim.Adam(self.parameters(), lr=self.lr, weight_decay=self.wd)
            elif self.optimizer == 'SGD':
                optimizer = torch.optim.SGD(self.parameters(), lr=self.lr, weight_decay=self.wd)
            elif self.optimizer == 'Adagrad':
                optimizer = torch.optim.Adagrad(self.parameters(), lr=self.lr, weight_decay=self.wd)
            return optimizer

**Comment by rohitgr7** [comment]
hey @dmandair !

the extra computation in `validation_epoch_end` you added later will be taking some memory and might be causing this memory issue. You can try computing these metrics at the batch level so that they can be computed efficiently for the complete epoch. You can try [torchmetrics](https://torchmetrics.readthedocs.io/en/latest/). The metrics you are using are already available there. This will help to avoid the memory allocation for all the logits and target values and you might not need to return anything from `validation_step` at all and `validation_epoch_end` could be removed completely.

**Reply by dmandair** [reply]
Thanks for the reply @rohitgr7 - interesting, I can definitely give that a shot! Quick follow up questions on how the code might look for that - 1. would I just flag to log the losses (both for training and validation) 'on_epoch' to only record these at the end of the epoch? 2. would I need to use things like loss.item or labels. data to avoid messing with gradients prior to logging this during the training_step (for when I record things like average loss, accuracy, etc.)? I wasn't sure how pytorch lightning handles detaching on the backend during training_steps. During the validation step, from what I understand I don't need to worry about this since the equivalent of 'with torch.no_grad()'. Would appreciate your thoughts on how I might write that code in practice. 

Thanks again for the help!

**Reply by rohitgr7** [reply]
if you want to record them only on epoch level, all you need to do
```py
def training_step(...)
    self.log('train_loss', loss, on_step=False, on_epoch=True)
    return loss

def validation_step(...):
    self.log('val_loss', loss, on_step=False, on_epoch=True)
```
note that setting `on_epoch=True` will aggregate the loss for the complete dataset and you don't need to take care of taking the mean in `_epoch_end` hooks. Check out logging docs for more info: https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html

just pass in `self.log('train_loss', loss, ...)`, don't worry about `.item/.detach`, it will be handled automatically.

yes, validation/testing/prediction runs under `no_grad`.

**Reply by dmandair** [reply]
awesome, thanks @rohitgr7! I'll give that a try and let you know how it goes. 

One other question that came up - what would be best practices to make sure no subprocesses are running on a GPU that pytorch lightening 'chooses'? I ask because as a part of my institution, we submit our GPU jobs to a scheduling system that assigns the node and technically also the GPU - from some test runs, I think if I run instantiate my pytorch lightening Trainer with (gpus = 1) with no other flags, it occasionally tries to run the script on a GPU on the node that has other subprocesses going on (and that contribute to the memory error). Are there pytorch lightening specific approaches to dealing with that (either removing subprocesses on a GPU or making sure the script runs on the right GPU)?

**Reply by rohitgr7** [reply]
I think you are looking for https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#auto-select-gpus

**Reply by ananthsub** [reply]
I strongly recommend using torchmetrics to keep metric state aggregated during the training step. I've personally found this to be more explicit and easier to rely on than the `on_step` and `on_epoch` arguments to LightningModule.log

**Reply by dmandair** [reply]
@ananthsub sorry maybe I'm not understanding but how do you do aggregation across batches if you're using torchmetrics within the training_step function (and not using the on_epoch flags to LighningModule.log)?

**Reply by ananthsub** [reply]
You call metric.update(...) to aggregate across batches, and metric.compute() whenever you want the result based on the metric states

**Reply by dmandair** [reply]
@ananthsub I guess I meant if you're aggregating with metric across batches, where is the metric instance instantiated? Wouldn't it have to be outside the 'training_step' (or validation step) function to keep the running total? Also, similar to using self.log, can we ignore things like loss.item() and label.data when using this during training? Thanks again!

**Reply by ananthsub** [reply]
You can instantiate the metric inside your LightningModule constructor, and use it across training and validation steps. There's more documentation here: https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html
